mail_browser_view/models/mail_mail.py

```diff
@@ -0,0 +1,105 @@
+from uuid import uuid4
+from base64 import urlsafe_b64encode, urlsafe_b64decode
+import binascii
+import lxml
+from werkzeug.urls import url_parse
+from datetime import date
+from dateutil.relativedelta import relativedelta
+
+
+class Mail(models.Model):
+    _inherit = 'mail.mail'
+
+    @api.model
+    def create(self, vals):
+        rec = super(Mail, self).create(vals)
+        rec._replace_view_url()
+        return rec
+
+    @api.model
+    def get_record_for_token(self, token):
+        try:
+            token = urlsafe_b64decode(token).decode()
+            access_token, rec_id = token[:32], token[32:]
+            rec = self.sudo().browse(int(rec_id))
+            return (rec.access_token == access_token and
+                    rec.is_token_alive and rec or self)
+        except (ValueError, MissingError, binascii.Error):
+            return self
+
+    @api.multi
+    def _replace_view_url(self):
+        self.ensure_one()
+
+        root_html = lxml.html.fromstring(self.body_html)
+        link_node = root_html.xpath("//a[@id='view_in_browser_url']")
+
+        if link_node:
+            base_url = self.env['ir.config_parameter'].sudo().get_param(
+                'web.base.url')
+            base = url_parse(base_url)
+
+            view_url = url_parse(
+                self.view_in_browser_url or '#'
+            ).replace(
+                scheme=base.scheme, netloc=base.netloc
+            ).to_url()
+
+            link_node[0].set('href', view_url)
+
+            self.body_html = lxml.html.tostring(
+                root_html,
+                pretty_print=False,
+                method='html',
+                encoding='unicode'
+            )
+
+    @api.depends('create_date')
+    def _compute_access_token(self):
+        for rec in self:
+            rec.access_token = uuid4().hex
+
+    @api.depends('access_token')
+    def _compute_browser_url(self):
+        for rec in self:
+            url_token = urlsafe_b64encode(
+                (rec.access_token + str(rec.id)).encode()
+            ).decode()
+            rec.view_in_browser_url = '/email/view/{}'.format(url_token)
+
+    @api.depends('mail_message_id',
+                 'mail_message_id.date')
+    def _compute_token_alive(self):
+        expiration_time = int(
+            self.env['ir.config_parameter'].sudo().get_param(
+            'mail_browser_view.token_expiration_hours') or '0'
+        )
+        if expiration_time > 0:
+            max_delta = relativedelta(hours=expiration_time)
+            for rec in self:
+                mail_date = fields.Datetime.from_string(
+                    rec.mail_message_id.date
+                )
+                rec.is_token_alive = (
+                    (mail_date + max_delta).date() >= date.today()
+                )
+        else:
+            self.is_token_alive = True
+
+    access_token = fields.Char(
+        'Security Token',
+        compute="_compute_access_token", store=True
```

Comment:
do you really need this to be a field and be stored? You are passing the id of the record in the url...
Either you rely on the ID and you use the token just to make some noise in the URL (no need to store it)
or you use the token to retrieve the message (no need to use the ID in the URL).

Reply:
fine, you can keep them to strength the uniqueness but then you use both to search the record ;)

Reply:
Currently, I do a browse with the ID and check if the access_token is matching if a record is found. Should I do a search with both values instead ? What's best ?

Reply:
search using both